llvm/docs/DirectX/DXILOpTableGenDesign.rst

```diff
@@ -172,14 +162,55 @@ Shader Model version is represented as follows:
      int Minor = minor;
    }
 
-   // Valid Shader model version records
+
+Concrete representations of valid Shader Model and DXIL versions are defined as follows:
+
+.. code-block::
 
    // Definition of Shader Model 6.0 - 6.8 and DXIL Version 1.0 - 1.8
    foreach i = 0...8 in {
      def SM6_#i : Version<6, i>;
      def DX1_#i : Version<1, i>;
    }
 
+Shader Stage Specification
+==========================
+
+Various shader stages such as ``compute``, ``pixel``, ``vertex``, etc., are represented as
+follows
+
+.. code-block::
+
+   // Shader stages
+   class ShaderStage;
+
+   def compute : ShaderStage;
+   def pixel : ShaderStage;
+   def vertex : ShaderStage;
+   ...
+
+Constraint Specification
+========================
+
+DXIL Operation properties such as valid overload types and valid shader stages are
+predicated on Shader Model version. These are represented as list of constraints.
+
+Following is the class representing a predicate and a constraint class representing
```

Comment:
I think it would be good to expand the description of how to interpret the `Constraint`. Is it that if the first predicate is true than the list of following predicates must be true?

And how do we interpret the list of constraints for the full dxil op?
`list<Constraints> constraints = [];`

Do we go through the list and find the first matching predicate? Do we find all matching predicates? Are the predicates allowed to overlap (i.e. can more than one predicate in the list be true).